contracts/fionft.sol

```diff
@@ -117,193 +94,196 @@ contract FIONFT is ERC721, ERC721Pausable {
           oracle_count--;
-          delete approvals[id];
-          uoracmapv++;
-          for(uint16 i = 0; i <= oraclelist.length - 1; i++) {
+          for(uint16 i = 0; i < oraclelist.length; i++) {
             if(oraclelist[i] == account) {
               oraclelist[i] = oraclelist[oraclelist.length - 1];
               oraclelist.pop();
               break;
             }
           }
-          emit oracle_unregistered(account, id);
+          emit oracle_unregistered(account, indexhash);
       }
+       emit consensus_activity("custodian", indexhash);
 
     } // unregoracle
 
-    function regcust(address account) external custodianOnly {
-      require(account != address(0), "Invalid address");
+    function regcust(address account) external onlyRole(CUSTODIAN_ROLE) {
+      require(account != address(0), "Invalid account");
       require(account != msg.sender, "Cannot register self");
-      bytes32 id = keccak256(bytes(abi.encode("rc",account, rcustmapv)));
-      require(!custodians[account].active, "Already registered");
-      require(!approvals[id].approved[msg.sender],  "Already approved");
-      int reqcust = custodian_count * 2 / 3 + 1;
-      if (approvals[id].approvals < reqcust) {
-        approvals[id].approvals++;
-        approvals[id].approved[msg.sender] = true;
-      }
-      if (approvals[id].approvals == reqcust) {
-        custodians[account].active = true;
+      bytes32 indexhash = keccak256(bytes(abi.encode(ApprovalType.AddCustodian,account)));
+      require(!hasRole(CUSTODIAN_ROLE, account), "Already registered");
+      if (getConsensus(indexhash, 1)) {
+        _grantRole(CUSTODIAN_ROLE, account);
         custodian_count++;
-        delete approvals[id];
-        rcustmapv++;
-        emit custodian_registered(account, id);
+        custodianlist.push(account);
+        emit custodian_registered(account, indexhash);
       }
+      emit consensus_activity("custodian", indexhash);
     }
 
-    function unregcust(address account) external custodianOnly {
-      require(account != address(0), "Invalid address");
-      require(custodians[account].active, "Custodian not registered");
-      require(custodian_count > 7, "Must contain 7 custodians");
-      bytes32 id = keccak256(bytes(abi.encode("uc",account, ucustmapv)));
-      require(!approvals[id].approved[msg.sender], "Already unregistered");
-      int reqcust = custodian_count * 2 / 3 + 1;
-      if (approvals[id].approvals < reqcust) {
-        approvals[id].approvals++;
-        approvals[id].approved[msg.sender] = true;
-      }
-      if ( approvals[id].approvals == reqcust) {
-          custodians[account].active = false;
-          delete custodians[account];
+    function unregcust(address account) external onlyRole(CUSTODIAN_ROLE) {
+      require(account != address(0), "Invalid account");
+      require(hasRole(CUSTODIAN_ROLE, account), "Custodian not registered");
+      require(custodian_count >= 7, "Must contain 7 custodians");
```

Comment:
We need to maintain a minimum of 7 custodians. If I am reading this correctly, if unregcust gets called and there are 7 custodians remaining, it will still succeed. I.e., unregcust should only be callable if there are 8+ custodians.